File: src/main/java/itmo/is/project/model/module/storage/StoredResource.java
package itmo.is.project.model.module.storage;

import itmo.is.project.model.resource.Resource;
import itmo.is.project.model.resource.ResourceAmount;
import itmo.is.project.model.resource.ResourceAmountHolder;
import jakarta.persistence.*;
import jakarta.validation.constraints.Min;
import jakarta.validation.constraints.NotNull;
import lombok.*;

@Entity
@Table(name = "stored_resource")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
public class StoredResource implements ResourceAmountHolder {

    @Embeddable
    public record CompositeKey(Integer storageModuleId, Integer resourceId) {
    }

    @EmbeddedId
    private StoredResource.CompositeKey id;

    @MapsId("storageModuleId")
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "storage_id", referencedColumnName = "id")
    private StorageModule storageModule;

    @MapsId("resourceId")
    @ManyToOne(fetch = FetchType.EAGER)
    @JoinColumn(name = "resource_id", referencedColumnName = "id")
    private Resource resource;

    @NotNull
    @Min(1)
    @Column(name = "amount", nullable = false)
    private Integer amount;

    public StoredResource(ResourceAmount resourceAmount, StorageModule storageModule) {
        setResourceAmount(resourceAmount);
        setStorageModule(storageModule);
        setId(new CompositeKey(storageModule.getId(), resource.getId()));
    }

    public void add(Integer amount) {
        this.amount += amount;
    }

    public void subtract(Integer amount) {
        this.amount -= amount;
    }
}

File: src/main/kotlin/itmo/isproject/model/module/storage/StoredResource.kt
package itmo.isproject.model.module.storage

import itmo.isproject.model.EmbeddedIdEntity
import itmo.isproject.model.resource.Resource
import itmo.isproject.model.resource.ResourceAmountHolder
import jakarta.persistence.*
import jakarta.validation.constraints.Min
import jakarta.validation.constraints.NotNull
import java.io.Serializable

@Entity
@Table(name = "stored_resource")
class StoredResource(

    @MapsId("storageModuleId")
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "storage_id", referencedColumnName = "id")
    var storageModule: StorageModule? = null,

    @MapsId("resourceId")
    @ManyToOne(fetch = FetchType.EAGER)
    @JoinColumn(name = "resource_id", referencedColumnName = "id")
    override var resource: Resource? = null,

    @NotNull
    @Min(1)
    @Column(name = "amount", nullable = false)
    override var amount: Int? = 0

) : EmbeddedIdEntity<StoredResourceId>(), ResourceAmountHolder {

    constructor(
        resourceAmount: ResourceAmountHolder,
        storageModule: StorageModule
    ) : this(
        resource = resourceAmount.resource,
        amount = resourceAmount.amount,
        storageModule = storageModule
    ) {
        id = StoredResourceId(storageModule.id, resource?.id)
    }
}

@Embeddable
data class StoredResourceId(
    val storageModuleId: Int?,
    val resourceId: Int?
) : Serializable